rholang/examples/linking/tests/LinkedList_test.rho

```diff
@@ -0,0 +1,289 @@
+        LinkedList <- LinkedListConstructor
+    ) {
+      new
+        emptyList_test, head_test, tail_test, prepend_test, fold_test,
+        reverse_test, foreach_test, map_test, get_test, indexOf_test,
+        length_test, range_test, LinkedList_test, and
+      in {
+        contract and(b1, b2, return) = {
+          match b1 with true => {
+            match b2 with true => {
+              return(true)
+            }
+            _ => { return(false) }
+          }
+          _ => { return(false) }
+        } |
+        contract emptyList_test(_) = {
+          //nothing much to test here, should just be the empty list constant
+          match emptyList == emptyList with true => { "EmptyList: Pass".display("\n") }
+          _ => { "EmptyList: Fail".display("\n") }
+        } |
+        contract head_test(_) = {
+          new NonEmptyList in {
+            LinkedList(1, NonEmptyList) | for(list <- NonEmptyList) {
+              new result1, result2, answer1, answer2 in {
+                head(list, result1) | head(emptyList, result2) |
+                answer1(1) | answer2(Nil) |
+                for(r1 <- result1; r2 <- result2; a1 <- answer1; a2 <- answer2) {
+                  match r1 == a1 with true => {
+                    match r2 == a2 with true => {
+                      "Head: Pass".display("\n")
+                    }
+                    _ => { "Head: Fail".display("\n") }
+                  }
+                  _ => { "Head: Fail".display("\n") }
+                }
+              }
+            }
+          }
+        } |
+        contract tail_test(_) = {
+          //return the tail of the list
+          new NonEmptyList in {
+            LinkedList(1, NonEmptyList) | for(tl <- NonEmptyList) {
+              prepend(2, tl, NonEmptyList) | for(list <- NonEmptyList) {
+                new result1, result2, answer1, answer2 in {
+                  tail(list, result1) | tail(emptyList, result2) |
+                  answer1(tl) | answer2(Nil) |
+                  for(r1 <- result1; r2 <- result2; a1 <- answer1; a2 <- answer2) {
+                    match r1 == a1 with true => {
+                      match r2 == a2 with true => {
+                        "Tail: Pass".display("\n")
+                      }
+                      _ => { "Tail: Fail".display("\n") }
+                    }
+                    _ => { "Tail: Fail".display("\n") }
+                  }
+                }
+              }
+            }
+          }
+        } |
+        contract prepend_test(_) = {
+          //create a new list with the supplied head attached
+          //to the supplied tail
+          new hd, tl, result in {
+            hd(1) | tl(emptyList) |
+            for(h <- hd; l <- tl) {
+              prepend(h, l, result) | for(r <- result) {
+                head(r, hd) | tail(r, tl) | 
+                for(rh <- hd; rl <- tl) {
+                  match rh == h with true => {
+                    match rl == l with true => {
```

Comment:
If you're using the same error message can we not match this as a tuple?

Reply:
There is a bunch of similar patterns down below.

Reply:
Yeah, good call